### PR TITLE
fix: prevent duplicate artifact assignment

### LINKED ISSUE #1242: Duplicate "Setting artifact for function" logs (possible redundant calls to `setArtifactPath`?)
# Duplicate "Setting artifact for function" logs (possible redundant calls to `setArtifactPath`?)

## Description

I guess this might be a minor bug but I did spend a fair chunk of time debugging it so I think it is worth raising. If the issue is valid I'll be happy to raise a PR to get a fix in.

### What went wrong?

Every time I package or deploy with the `--verbose` switch I get duplicate "Setting artifact for function" messages.

For example with individual packaging enabled and a function called "myFunction"

```
yarn sls package --stage sandbox --verbose
```

```
...
[Webpack] Copying existing artifacts
Setting artifact for function 'myFunction' to '.serverless/myFunction.zip'
Setting artifact for function 'myFunction2' to '.serverless/myFunction2.zip'
Setting artifact for function 'myFunction' to '.serverless/myFunction.zip'
Setting artifact for function 'myFunction2' to '.serverless/myFunction2.zip'
...
```

### What did you expect should have happened?

Not to have duplicate logs for each function

### What was the config you used?

```typescript

import type { AWS } from '@serverless/typescript';
import myFunction from '@functions/myFunction';
import myFunction2 from '@functions/myFunction2';

const serverlessConfiguration: AWS = {
  service: 'example-project',
  frameworkVersion: '3',
  plugins: ['serverless-webpack'],
  provider: {
    name: 'aws',
    region: 'ap-southeast-2',
    runtime: 'nodejs14.x',
    stage: '${opt:stage}',
    apiGateway: {
      minimumCompressionSize: 1024,
      shouldStartNameWithService: true,
      apiKeys: ['foo-bar'],
    },
    environment: {
      AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
      NODE_OPTIONS: '--enable-source-maps --stack-trace-limit=1000',
    },
  },
  // import the function via paths
  functions: {
    myFunction,
    myFunction2,
  },
  package: { individually: true },
  resources: {
    Resources: {},
  },
  custom: {
    webpack: {
      packager: 'yarn',
      includeModules: {
        forceExclude: ['some-lib'],
      },
    },
  },
};

module.exports = serverlessConfiguration;

```

### What stacktrace or error message from your provider did you see?

No explicit stack trace or error message

## Suggested Fix

I could be wrong but it looks like there might be some duplicate code in `lib/packageModules.js`

First there is a loop to call `setArtifactPath` for each function.
https://github.com/serverless-heaven/serverless-webpack/blob/76988d5f70cb2d8e61591d9188ee0c4d0872ccec/lib/packageModules.js#L236

it already has logic to set the correct artefact path depending on whether individual packaging is enabled or not

https://github.com/serverless-heaven/serverless-webpack/blob/76988d5f70cb2d8e61591d9188ee0c4d0872ccec/lib/packageModules.js#L239

Then further down, if individual packaging is enabled, `setArtifactPath` is called again with the same args which I think is redundant

https://github.com/serverless-heaven/serverless-webpack/blob/76988d5f70cb2d8e61591d9188ee0c4d0872ccec/lib/packageModules.js#L247

If I understand correctlty, I think the following patch might do the trick, without introducing any side effects. It is very rough, and I haven't tested it. If you think my analysis is correct I'll go over it again and make sure tests pass etc.

See what you think:

```patch
---
 lib/packageModules.js | 23 +++++++----------------
 1 file changed, 7 insertions(+), 16 deletions(-)

diff --git a/lib/packageModules.js b/lib/packageModules.js
index fce5493..e794b29 100644
--- a/lib/packageModules.js
+++ b/lib/packageModules.js
@@ -224,42 +224,33 @@ module.exports = {
     // only function passed with `-f` flag should be processed.
     const functionNames = this.options.function ? [this.options.function] : getAllNodeFunctions.call(this);
 
+    const serviceName = this.serverless.service.getServiceObject().name;
+
     // Copy artifacts to package location
     if (isIndividualPackaging.call(this)) {
       _.forEach(functionNames, funcName => copyArtifactByName.call(this, funcName));
     } else {
       // Copy service packaged artifact
-      const serviceName = this.serverless.service.getServiceObject().name;
       copyArtifactByName.call(this, serviceName);
     }
 
     _.forEach(functionNames, funcName => {
       const func = this.serverless.service.getFunction(funcName);
 
-      const archiveName = isIndividualPackaging.call(this) ? funcName : this.serverless.service.getServiceObject().name;
+      const archiveName = isIndividualPackaging.call(this) ? funcName : serviceName;
 
       const { serverlessArtifact } = getArtifactLocations.call(this, archiveName);
       setArtifactPath.call(this, funcName, func, serverlessArtifact);
     });
 
     // Set artifact locations
-    if (isIndividualPackaging.call(this)) {
-      _.forEach(functionNames, funcName => {
-        const func = this.serverless.service.getFunction(funcName);
-
-        const archiveName = funcName;
-
-        const { serverlessArtifact } = getArtifactLocations.call(this, archiveName);
-        setArtifactPath.call(this, funcName, func, serverlessArtifact);
-      });
-    } else {
-      const archiveName = this.serverless.service.getServiceObject().name;
+    if (!isIndividualPackaging.call(this) && 
+        _.get(this.serverless, 'service.provider.name') === 'google') {
+      const archiveName = serviceName;
 
       const { serverlessArtifact } = getArtifactLocations.call(this, archiveName);
 
-      if (_.get(this.serverless, 'service.provider.name') === 'google') {
-        setServiceArtifactPath.call(this, serverlessArtifact);
-      }
+      setServiceArtifactPath.call(this, serverlessArtifact);
     }
 
     return BbPromise.resolve();
-- 
```

## Additional Data

* ***Serverless-Webpack Version you're using***:
5.7.1
* ***Webpack version you're using***:
5.73.0
* ***Serverless Framework Version you're using***:
Framework Core: 3.7.1 (local)
Plugin: 6.1.5
SDK: 4.3.2
* ***Operating System***:
MacOS Monterey 12.6

## Also

Thanks for the great plugin!

### PR DESCRIPTION
**Pull request by NoxHarmonium**
## What did you implement:

Prevent duplicate artifact assignment, extracted some duplicate logic.

Closes #1242

## How did you implement it:

Some minor refactoring so that code paths are not called twice redundantly (see issue #1242)

Old Coverage:

```

=============================== Coverage summary ===============================
Statements   : 95.92% ( 2566/2675 )
Branches     : 89.67% ( 530/591 )
Functions    : 98.18% ( 108/110 )
Lines        : 95.92% ( 2566/2675 )
================================================================================
```

New Coverage:

```
=============================== Coverage summary ===============================
Statements   : 95.95% ( 2584/2693 )
Branches     : 89.71% ( 532/593 )
Functions    : 98.19% ( 109/111 )
Lines        : 95.95% ( 2584/2693 )
================================================================================
```

## How can we verify it:

1. Run `sls package --verbose` with individual packaging switched on.
2. Grep for "Setting artifact for function"
3. Check that each function is only mentioned once

I have tested this in my local project and it works for me.

## Todos:

- [X] Write tests
- [X] Write documentation
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/packageModules.js [hunk 1]
```diff
@@ -223,43 +223,40 @@ module.exports = {
...
-        setArtifactPath.call(this, funcName, func, serverlessArtifact);
-      });
-    } else {
-      const archiveName = this.serverless.service.getServiceObject().name;
+    // If we are deploying to 'google' we need to set an artifact for the whole service,
+    // rather than for each function, so there is special case here
+    if (!individualPackagingEnabled && providerIsGoogle) {
+      const archiveName = serviceName;
 
+      // Ths may look similar to the loop above, but note that this calls
```
Comment (j0k3r) [inline lib/packageModules.js:256]:
```suggestion
      // This may look similar to the loop above, but note that this calls
```

**Review by j0k3r — COMMENTED** [review]
Good job, just found a small typo 👍🏼

**Comment by NoxHarmonium** [comment]
Typo fixed and branch rebased!

**Review by vicary — APPROVED** [review]
LGTM

**Review by j0k3r — APPROVED** [review]

**Pull request merged**